### PR TITLE
[HEI-269] Twin search pipline 

### PR DESCRIPTION
**Pull request by VasilievaKA**
test = TWIN_SEARCH
result = test.execute(ExperimentData(data))

### CODE REVIEW

File: hypex/comparators/abstract.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def _get_fields(self, data: ExperimentData):
     def _execute_inner_function(
         cls,
         grouping_data,
-        target_fields: Optional[List[FieldKeyTypes]] = None,
+        target_field: Optional[List[FieldKeyTypes]] = None,
```
Comment (Dmatryus) [inline hypex/comparators/abstract.py:62]: If the `List` is still `fields`.

File: hypex/comparators/distances.py [hunk 2]
```diff
@@ -1,19 +1,75 @@
...
+    @classmethod
+    def _execute_inner_function(
+        cls,
+        grouping_data,
+        target_field: Optional[List[FieldKeyTypes]] = None,
+        **kwargs,
+    ) -> Dict:
+        result = {}
+        for i in range(1, len(grouping_data)):
+            if target_field:
```
Comment (Dmatryus) [inline hypex/comparators/distances.py:20]: It's better to do something like this:
```
t_data = grouping_data[0][1][target_field] if target_field else grouping_data[0][1]
t_test_data = grouping_data[i][1][target_field] if target_field else grouping_data[i][1]
result.update(
    cls._inner_function(
        data=t_data,
        test_data=t_test_data,
        **kwargs,
    )
)
```

File: hypex/executor/executor.py [hunk 3]
```diff
@@ -243,7 +237,142 @@ def execute(self, data: ExperimentData) -> ExperimentData:
...
             grouping_data=grouping_data,
+            target_field=target_fields,
+        )
+        return self._set_value(data, compare_result)
+
+
+class MLExecutor(GroupCalculator, ABC):
+    def __init__(
+        self,
+        grouping_role: Optional[ABCRole] = None,
```
Comment (Dmatryus) [inline hypex/executor/executor.py:249]: GroupingRole() and target_role by default?

File: hypex/executor/executor.py [hunk 4]
```diff
@@ -243,7 +237,142 @@ def execute(self, data: ExperimentData) -> ExperimentData:
...
+
+    def _get_fields(self, data: ExperimentData):
+        group_field = self._field_searching(data, self.grouping_role)
+        target_field = self._field_searching(
+            data, self.target_role, search_types=self.search_types
+        )
+        return group_field, target_field
+
+    @abstractmethod
+    def fit(self, X: Dataset, Y: Dataset) -> "MLExecutor":
```
Comment (Dmatryus) [inline hypex/executor/executor.py:265]: Y -> Optional for Unsupervised methods

File: hypex/executor/executor.py [hunk 5]
```diff
@@ -243,7 +237,142 @@ def execute(self, data: ExperimentData) -> ExperimentData:
...
+        raise NotImplementedError
+
+    @abstractmethod
+    def predict(self, data: Dataset) -> Dataset:
```
Comment (Dmatryus) [inline hypex/executor/executor.py:269]: `X` instead of `data` for uniformity

File: hypex/executor/executor.py [hunk 6]
```diff
@@ -243,7 +237,142 @@ def execute(self, data: ExperimentData) -> ExperimentData:
...
+        raise NotImplementedError
+
+    def score(self, X: Dataset, Y: Dataset) -> float:
```
Comment (Dmatryus) [inline hypex/executor/executor.py:272]: `score` is usually calculated from 2 `Y`

File: hypex/extensions/abstract.py [hunk 7]
```diff
@@ -23,12 +21,46 @@ def _calc_pandas(self, data: Dataset, **kwargs):
...
+        target_data: Optional[Dataset] = None,
+        test_data: Optional[Dataset] = None,
+        mode: Optional[Literal["auto", "fit", "predict"]] = None,
+        **kwargs
+    ):
+        model = self.fit(data, target_data, **kwargs)
+        return model.predict(test_data)
+
+    @abstractmethod
+    def fit(self, X, y=None, **kwargs):
```
Comment (Dmatryus) [inline hypex/extensions/abstract.py:50]: `Y` for uniformity

File: hypex/extensions/abstract.py [hunk 8]
```diff
@@ -23,12 +21,46 @@ def _calc_pandas(self, data: Dataset, **kwargs):
...
+
+class MLExtension(Extension):
+    def _calc_pandas(
+        self,
+        data: Dataset,
+        target_data: Optional[Dataset] = None,
+        test_data: Optional[Dataset] = None,
+        mode: Optional[Literal["auto", "fit", "predict"]] = None,
+        **kwargs
+    ):
```
Comment (Dmatryus) [inline hypex/extensions/abstract.py:45]: `mode` is not used

File: hypex/extensions/ml.py
Comment (Dmatryus) [inline hypex/extensions/ml.py:1]: `extensions/faiss` maybe?

File: hypex/ml/faiss.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    def predict(self, data: Dataset) -> Dataset:
+        return FaissExtension().predict(data)
+
+    def execute(self, data: ExperimentData) -> ExperimentData:
+        group_field, features_fields = self._get_fields(data=data)
+        if group_field[0] in data.groups:
+            grouping_data = list(data.groups[group_field[0]].items())
+        else:
+            grouping_data = None
+        distances_keys = data.get_ids(MahalanobisDistance, ExperimentDataEnum.groups)
```
Comment (Dmatryus) [inline hypex/ml/faiss.py:73]: Distance may be not `Mahalanobis`

File: hypex/operators/abstract.py [hunk 10]
```diff
@@ -24,28 +24,29 @@ def _inner_function(
...
-    def _get_fields(self, data: ExperimentData): 
+
+    def _get_fields(self, data: ExperimentData):
         group_field = self._field_searching(data, self.grouping_role)
         target_fields = self._field_searching(
             data, self.target_roles, search_types=self.search_types
         )
         return group_field, target_fields
 
+    # TODO дописать ошибку
```
Comment (Dmatryus) [inline hypex/operators/abstract.py:35]: TODO, TODO, todo, todo, todo, todo, todo, tototototodo...

274px-Розовая_пантера_(персонаж)

**Review by Dmatryus — COMMENTED** [review]
AAAC8ZOkOSOT9rnpiX6yPF8z9awCTRiNWJ9OxTGYbtqlJKixGn4tEg_DTkVWlmKjkKnOH5qc11q6xKFRF5VLlwgA-DA

**Review by TonyKatkov89 — APPROVED** [review]
Exceptional, fantastic, superb, awsome, terrific, stunning!

File: hypex/executor/executor.py [hunk 11]
```diff
@@ -182,7 +177,7 @@ def _field_arg_universalization(
     ) -> List[FieldKeyTypes]:
         if not field:
             raise NoColumnsError(field)
-        elif isinstance(field, FieldKeyTypes):
+        elif isinstance(field, int) or isinstance(field, str):
```
Comment (Dmatryus) [inline hypex/executor/executor.py:180]: Only `isinstance(field, str)`

File: hypex/operators/abstract.py [hunk 12]
```diff
@@ -24,23 +24,24 @@ def _inner_function(
...
     def _execute_inner_function(
         cls,
         grouping_data,
         target_fields: Optional[List[FieldKeyTypes]] = None,
         **kwargs,
     ) -> Dict:
-        if len(target_fields) != 2:
-            raise ValueError
+        if target_fields is None or len(target_fields) != 2:
+            raise ValueError("Нужно дописать ошибку")
```
Comment (Dmatryus) [inline hypex/operators/abstract.py:44]: ?

**Review by Dmatryus — COMMENTED** [review]
Not everything is smooth and clear.

File: hypex/comparators/distances.py [hunk 13]
```diff
@@ -1,19 +1,75 @@
...
+                result.update(
+                    cls._inner_function(
+                        data=grouping_data[0][1],
+                        test_data=grouping_data[i][1],
+                        **kwargs,
+                    )
+                )
+        return result
+
+    def _set_value(
```
[the remark this answers is not in the document]
Comment (TonyKatkov89) [reply hypex/comparators/distances.py:37]: still valid

**Review by TonyKatkov89 — COMMENTED** [review]
╰(*°▽°*)╯

**Pull request merged**